inventory/s3/orc_reader.go

```diff
@@ -0,0 +1,155 @@
+package s3
+
+import (
+	"context"
+	"os"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/go-openapi/swag"
+	"github.com/hashicorp/go-multierror"
+	"github.com/scritchley/orc"
+	"modernc.org/mathutil"
+)
+
+type OrcInventoryFileReader struct {
+	reader                   *orc.Reader
+	c                        *orc.Cursor
+	ctx                      context.Context
+	selectIndexByField       map[string]int
+	actualColumnIndexByField map[string]int
+	mgr                      *Reader
+	key                      string
+}
+
+type orcFile struct {
+	key           string
+	localFilename string
+	ready         bool
+}
+
+func getSelectFields(typeDescription *orc.TypeDescription) ([]string, map[string]int, map[string]int) {
+	relevantFields := []string{"bucket", "key", "size", "last_modified_date", "e_tag", "is_delete_marker", "is_latest"}
+	actualColumnIndexByField := make(map[string]int)
+	selectFields := make([]string, 0, len(typeDescription.Columns()))
+	selectIndexByField := make(map[string]int)
+	for i, field := range typeDescription.Columns() {
+		actualColumnIndexByField[field] = i
+	}
+	j := 0
+	for _, field := range relevantFields {
+		if _, ok := actualColumnIndexByField[field]; ok {
+			selectFields = append(selectFields, field)
+			selectIndexByField[field] = j
+			j++
+		}
+	}
+	return selectFields, selectIndexByField, actualColumnIndexByField
+}
+
+func (r *OrcInventoryFileReader) inventoryObjectFromRow(rowData []interface{}) InventoryObject {
+	var size *int64
+	if sizeIdx, ok := r.selectIndexByField["size"]; ok && rowData[sizeIdx] != nil {
+		size = swag.Int64(rowData[sizeIdx].(int64))
+	}
+	var lastModified *int64
+	if lastModifiedIdx, ok := r.selectIndexByField["last_modified_date"]; ok && rowData[lastModifiedIdx] != nil {
+		lastModified = swag.Int64(rowData[lastModifiedIdx].(time.Time).Unix())
+	}
+	var eTag *string
+	if eTagIdx, ok := r.selectIndexByField["e_tag"]; ok && rowData[eTagIdx] != nil {
+		eTag = swag.String(rowData[eTagIdx].(string))
+	}
+	var isLatest *bool
+	if isLatestIdx, ok := r.selectIndexByField["is_latest"]; ok && rowData[isLatestIdx] != nil {
+		isLatest = swag.Bool(rowData[isLatestIdx].(bool))
+	}
+	var isDeleteMarker *bool
+	if isDeleteMarkerIdx, ok := r.selectIndexByField["is_delete_marker"]; ok && rowData[isDeleteMarkerIdx] != nil {
+		isDeleteMarker = swag.Bool(rowData[isDeleteMarkerIdx].(bool))
+	}
+	return InventoryObject{
+		Bucket:         rowData[r.selectIndexByField["bucket"]].(string),
+		Key:            rowData[r.selectIndexByField["key"]].(string),
+		Size:           size,
+		LastModified:   lastModified,
+		Checksum:       eTag,
+		IsLatest:       isLatest,
+		IsDeleteMarker: isDeleteMarker,
+	}
+}
+
+func (r *OrcInventoryFileReader) Read(dstInterface interface{}) error {
+	num := reflect.ValueOf(dstInterface).Elem().Len()
+	res := make([]InventoryObject, 0, num)
```

Comment:
This is not a polymorphic call.  AFAICT it is only safe to call `OIFR.Read` and pass an argument of type `InventoryObject`.  Meanwhile presumably it is only safe to call `PrqR.Read` and pass an argument of type `prq.SomeObject`.  So it is _not_ part of the interface: it is only safe to call `reader.Read` if you already know the type.

I would suggest the following instead.  Remove `Read` from the interface.  Now calling code has to say e.g.
```go
if oifr, ok := reader.(*OrcInventoryFileReader); ok {
        invObs := make([]InventoryObject, 0)
        err = oifr.Read(&invObs)
        // ...
} else if prqr, ok := reader.(*ParquetSomethingReader); ok {
        prqObs := make([]ParquetInventoryObject, 0)
        err = prqr.Read(&prqObs)
        // ...
}
```
and everyone's happy, even the Go compiler, and _I don't need to think about reflection and complain about Go_ (which is my true goal here).

Reply:
I would still prefer code that does a dynamic cast (as above) and avoids reflection.  We're at something of an impasse here; maybe we should just ask someone on the team with mad Golang skillz and ask whether this use of reflection seems reasonable...